### ISSUE TITLE
moncular depth image visualization problem

### ISSUE THREAD

**Issue by LevinJ** [OPEN-AIR-SUN/mars#92]
When training with mono depth supervision enabled, the background depth visualization does not look right as below

![ko_bg](https://github.com/OPEN-AIR-SUN/mars/assets/1629014/00920feb-8742-4003-aa47-e9adb6f19dfb)

Looking into the `get_image_metrics_and_images` method implementation in scene_graph.py, I noticed  below codes,

```
max_depth = depth_gt.max()
....
depth[depth > max_depth] = max_depth  # depth always has some very far value
....
background_depth[background_depth > max_depth] = max_depth
```
In the case of monocular depth, ground truth `gt_depth` and predicted depth `depth` and  `background_depth` are not necessarily of the same scale, clipping predicted depth by a fixed max value might incorrectly set  all predicted depth values to a single value (for example, when gt_depth is in the range of 0 and 1, and all predicted depth are larger than 1 ).

After removing clipping operation as below,  the background depth visualization looks ok.

```
if self.use_depth_loss:
      # align to predicted depth and normalize
      background_depth = outputs["background_depth"]
      depth_gt = batch["depth_image"].to(depth)
      depth_mask = batch["depth_mask"].to(self.device)
      if not self.config.is_euclidean_depth:
          depth_gt = depth_gt * outputs["directions_norm"]
      depth_gt[~depth_mask] = 0.0  # to make it compatible with the automask of the depth loss
      depth[~depth_mask] = 0.0
      

      if self.config.mono_depth_loss_mult > 1e-8:
          # align to predicted depth and normalize
          scale, shift = normalized_depth_scale_and_shift(
              depth[None, ...], depth_gt[None, ...], depth_mask[None, ...]
          )
          depth = depth * scale + shift
          depth[~depth_mask] = 0.0 # set masked area back to zero after adding the shift
      else:
          max_depth = depth_gt.max()
          depth[depth > max_depth] = max_depth  # depth always has some very far value
          background_depth[background_depth > max_depth] = max_depth
      
      background_depth[~depth_mask] = 0.0`
```

![ok_bg](https://github.com/OPEN-AIR-SUN/mars/assets/1629014/0da1e67a-6880-4d5a-a3d2-c56d0e077b61)

**Comment by wuzirui** [comment]
Hi @LevinJ, thank you for reaching out. We will consider making the clipping threshold an adjustable config in the next version (PRs are welcomed btw hah).

**Comment by LevinJ** [comment]
Thanks for the quick feedback!  I am not very sure about what you mean by "making the clipping threshold an adjustable config".
From what I understand, to fix the problem, we could simply skip clipping operation for monocular depth visualization. In more details:

1. For the monocular depth supervision visualization, ground truth depth and predicted truth are not of identical scale, clipping operation probably should not be applied.
2. For sensor depth supervision visualization, ground truth depth and predicted truth are expected to be of same scale, we could keep the current clipping operation to have a nicer visualization effect.

Once I am clear about your intent on how to fix the issue, I shall be very happy to create a PR.

**Comment by wuzirui** [comment]
Hi, thank you for raising this concern. The reason I would insist on using a clipping threshold is that for outdoor scenes, there exist regions that are far from the camera viewpoints (e.g., the sky). In such cases, the depth variance when visualizing depth maps on near regions is small if we do not apply depth clipping (because all the depth values are scaled into [0,1] to apply colormaps).

**Comment by LevinJ** [comment]
Good analysis. yes, I see your points about depth variance being small in the near region if we don't clipping the sky region.

Lately I've been trying adding moncular depth priors supervision to the pipeline (using vkitti dataset, and extracting monocular depth by omindata), without much success. The problems I observed include: 
1) predicted depth is noisy and does not match well with ground truth (especially in the sky region), 
2) the depth loss jump around and does not decrease
3) don't see much improvement in PSNR after adding monocular depth supervision

This leads me to think that maybe the ground truth depth for sky region is not reliable in the first place, thus causing above problems. So maybe one option is to use a depth mask to mask out the sky region and prevent them from contributing to the depth loss supervision.  During the visualization stage, we could accordingly mask out the sky region for both ground truth depth and predicted depth.

**Comment by wuzirui** [comment]
Yeah, I totally agree. Actually, we have tried using the `sky_mask` to separate the original depth loss into two parts: one minimizes the depth difference between gt and pred at the non-sky region, and one minimizes the disparity (1 / depth) at the sky region. It turns out to be slightly better than using the mono-depth loss. Maybe we can add it in some future version.

**Comment by LevinJ** [comment]
Interesting idea about minimizing the disparity for sky region. I guess the motivation is, for the sky region, its disparity is small, as a result, minimizing disparity gap between gt and prediction would  would actually assigning a small weight to the sky region depth loss. Is this the case?

I am truly curious about how your mono depth trainings go. If it's convenient, can you help post your  mono depth training loss trend chart? Thanks.

**Comment by wuzirui** [comment]
In fact, for the sky region, our new depth loss is not minimizing the gap, but minimizing the predicted disparity. Maybe I will find some wandb report and get back to you soon~

**Comment by xBeho1der** [comment]
Hi! @LevinJ 

Here is a [wandb report](https://api.wandb.ai/links/wuzirui-research/m3d471zj) link with `far_depth_loss` in `Train Loss Dict`.

**Comment by LevinJ** [comment]
Hi @xBeho1der , thanks a lot for the wand charts. Looking at your depth_loss in Train Loss Dict below, I tend to think that your training mono depth loss is converging well.

![depth_loss](https://github.com/OPEN-AIR-SUN/mars/assets/1629014/bf4f472e-2fc4-4bf7-a8ed-99c70d6dab30)

Lately I've been trying  mono depth supervision on kitti dataset with no success, the mono depth training loss does not  really decrease. If relevant codes for reproducing your above wandb report is open sourced, I would appreciate very much if you could help and share  the training config so that I could find the root cause of my failure.

**Comment by xBeho1der** [comment]
Hi! @LevinJ 

For monocular depth training, you may refering the following config:

```python
            model=SceneGraphModelConfig(
                background_model=NerfactoModelConfig(),
                object_model_template=CarNeRFModelConfig(_target=CarNeRF),
                object_representation="class-wise",
                object_ray_sample_strategy="remove-bg",
                mono_depth_loss_mult=0.01,
                depth_loss_mult=0,
            ),
 ```

**Comment by LevinJ** [comment]
Got it. Thanks a lot for the information.
